src/main/java/com/github/mtsochacki/adventofcode/Day18.java

```diff
@@ -0,0 +1,183 @@
+                }
+                List<RegularNumber> firstSlug = copySnailfishNumber(snailfishNumbers.get(i));
+                List<RegularNumber> copyFirstSlug = copySnailfishNumber(snailfishNumbers.get(i));
+                List<RegularNumber> secondSlug = copySnailfishNumber(snailfishNumbers.get(j));
+                List<RegularNumber> copySecondSlug = copySnailfishNumber(snailfishNumbers.get(j));
+                List<RegularNumber> firstResult = reduceSnailfishNumber(firstSlug, secondSlug);
+                List<RegularNumber> secondResult = reduceSnailfishNumber(copySecondSlug, copyFirstSlug);
+                int newMax = Math.max(calculateMagnitude(firstResult), calculateMagnitude(secondResult));
+                max = Math.max(newMax, max);
+            }
+        }
+        return max;
+    }
+
+    private List<RegularNumber> copySnailfishNumber(List<RegularNumber> snailfishNumber) {
+        List<RegularNumber> newSnailfishNumber = new ArrayList<>();
+        for (RegularNumber regularNumber : snailfishNumber) {
+            newSnailfishNumber.add(new RegularNumber(regularNumber));
+        }
+        return newSnailfishNumber;
+    }
+
+    private List<RegularNumber> transformIntoSnailfishNumber(String line) {
+        List<RegularNumber> snailfishNumber = new ArrayList<>();
+        int level = 0;
+        for (char character : line.toCharArray()) {
+            if (character == '[') {
+                level++;
+            } else if (character == ']') {
+                level--;
+            } else if (Character.isDigit(character)) {
+                snailfishNumber.add(new RegularNumber(level, Character.getNumericValue(character)));
+            }
+        }
+        return snailfishNumber;
+    }
+
+    private List<RegularNumber> reduceSnailfishNumber(List<RegularNumber> firstSnailfish, List<RegularNumber> secondSnailfish) {
+        List<RegularNumber> resultSlug = addSnailfish(firstSnailfish, secondSnailfish);
+        return reduceSlug(resultSlug);
+    }
+
+    private List<RegularNumber> addSnailfish(List<RegularNumber> firstSnailfish, List<RegularNumber> secondSnailfish) {
+        firstSnailfish.addAll(secondSnailfish);
+        for (RegularNumber slug : firstSnailfish) {
+            slug.incrementLevel();
+        }
+        return firstSnailfish;
+    }
+
+    private List<RegularNumber> reduceSlug(List<RegularNumber> snailfishNumber) {
+        List<RegularNumber> reducedSnailfish = copySnailfishNumber(snailfishNumber);
+        while (isInNeedOfExplosion(reducedSnailfish) || isInNeedOfSplit(reducedSnailfish)) {
+            if (isInNeedOfExplosion(reducedSnailfish)) {
+                explode(reducedSnailfish);
+            } else if (isInNeedOfSplit(reducedSnailfish)) {
+                split(reducedSnailfish);
+            }
+        }
+        return reducedSnailfish;
+    }
+
+    private boolean isInNeedOfExplosion(List<RegularNumber> snailfishNumber) {
+        return snailfishNumber.stream()
```

Comment:
Time saving idea: return the index at which the level is > 4, this way you don't have to iterate in `explode`

Reply:
done